packages/thyra/adapters/tpetra/src/Thyra_TpetraVectorSpace_def.hpp

```diff
@@ -114,6 +114,107 @@ TpetraVectorSpace<Scalar,LocalOrdinal,GlobalOrdinal,Node>::createMembers(int num
+TpetraVectorSpace<Scalar,LocalOrdinal,GlobalOrdinal,Node>::createMembersView(
+  const RTOpPack::SubMultiVectorView<Scalar> &raw_mv ) const
+{
+#ifdef TEUCHOS_DEBUG
+  TEUCHOS_TEST_FOR_EXCEPT( raw_mv.subDim() != this->dim() );
+#endif
+
+  // Create a multi-vector
+  RCP< MultiVectorBase<Scalar> > mv;
+  if (!tpetraMap_->isDistributed()) {
+
+    if (tpetraMV_.is_null() || (tpetraMV_->getNumVectors() != (size_t)raw_mv.numSubCols())) {
+      using IST = typename Tpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node>::impl_scalar_type;
+      using DT = typename Tpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node>::device_type;
+      auto dv = ::Tpetra::Details::getStatic2dDualView<IST, DT> (tpetraMap_->getGlobalNumElements(), raw_mv.numSubCols());
+      tpetraMV_ = Teuchos::rcp(new Tpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node>(tpetraMap_, dv));
+    }
+
+    if (tpetraDomainSpace_.is_null() || raw_mv.numSubCols() != tpetraDomainSpace_->localSubDim())
+      tpetraDomainSpace_ = tpetraVectorSpace<Scalar>(Tpetra::createLocalMapWithNode<LocalOrdinal, GlobalOrdinal, Node>(raw_mv.numSubCols(), tpetraMap_->getComm()));
+
+    mv = tpetraMultiVector<Scalar>(weakSelfPtr_.create_strong().getConst(), tpetraDomainSpace_, tpetraMV_);
+  } else {
+    mv = this->createMembers(raw_mv.numSubCols());
+  }
+  // Copy initial values in raw_mv into multi-vector
+  RTOpPack::SubMultiVectorView<Scalar> smv;
+  mv->acquireDetachedView(Range1D(),Range1D(),&smv);
+  RTOpPack::assign_entries<Scalar>(
+    Ptr<const RTOpPack::SubMultiVectorView<Scalar> >(Teuchos::outArg(smv)),
+    raw_mv
+    );
+  mv->commitDetachedView(&smv);
+  // Setup smart pointer to multi-vector to copy view back out just before multi-vector is destroyed
+  Teuchos::set_extra_data(
+    Teuchos::rcp(new CopyMultiVectorViewBack<Scalar>(&*mv,raw_mv)),
+    "CopyMultiVectorViewBack",
+    Teuchos::outArg(mv),
+    Teuchos::PRE_DESTROY
+    );
+  return mv;
+}
+
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+RCP<const MultiVectorBase<Scalar> >
+TpetraVectorSpace<Scalar,LocalOrdinal,GlobalOrdinal,Node>::createMembersView(
+  const RTOpPack::ConstSubMultiVectorView<Scalar> &raw_mv ) const
+{
+#ifdef TEUCHOS_DEBUG
+  TEUCHOS_TEST_FOR_EXCEPT( raw_mv.subDim() != this->dim() );
+#endif
+  // Create a multi-vector
+  RCP< MultiVectorBase<Scalar> > mv;
+  if (!tpetraMap_->isDistributed()) {
+    if (tpetraMV_.is_null() || (tpetraMV_->getNumVectors() != (size_t)raw_mv.numSubCols())) {
+      using IST = typename Tpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node>::impl_scalar_type;
+      using DT = typename Tpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node>::device_type;
+      auto dv = ::Tpetra::Details::getStatic2dDualView<IST, DT> (tpetraMap_->getGlobalNumElements(), raw_mv.numSubCols());
+      tpetraMV_ = Teuchos::rcp(new Tpetra::MultiVector<Scalar,LocalOrdinal,GlobalOrdinal,Node>(tpetraMap_, dv));
```

Comment:
Suppose that you create two different locally replicated multivectors with this function, and the second multivector has more columns.  The second call to `getStatic2dDualView` may then reallocate.  This will invalidate the first multivector's storage.  Is that OK here?  If not, then we need to write a proper memory pool and not just use this static thing.

Reply:
It's only ok as long as we don't try to use two of these temporary MV objects at the same time. Currently, we are not doing that anywhere. I could add a status variable to the MV that is checked before a `CopyMultiVectorViewBack` object is built, and reset once the copy-out happens. That way, we could throw, in case anyone ever tries to do this.

Reply:
@cgcgcg wrote:
> I could add a status variable to the MV that is checked before a `CopyMultiVectorViewBack` object is built, and reset once the copy-out happens. That way, we could throw, in case anyone ever tries to do this.

That would be super useful -- if you have time, could you please do that?  Thanks! :-D